fastcrypto/src/polynomial.rs

```diff
@@ -0,0 +1,185 @@
+pub struct Poly<C>(Vec<C>);
+
+pub type PrivatePoly<C> = Poly<<C as GroupElement>::ScalarType>;
+pub type PublicPoly<C> = Poly<C>;
+
+//// Vector related operations.
+
+impl<C> Poly<C> {
+    /// Returns the degree of the polynomial
+    pub fn degree(&self) -> usize {
+        // e.g. c_0 + c_1 * x + c_2 * x^2 + c_3 * x^3
+        // ^ 4 coefficients correspond to a 3rd degree poly
+        self.0.len() - 1
+    }
+}
+
+impl<C> From<Vec<C>> for Poly<C> {
+    fn from(c: Vec<C>) -> Self {
+        Self(c)
+    }
+}
+
+impl<C> From<Poly<C>> for Vec<C> {
+    fn from(poly: Poly<C>) -> Self {
+        poly.0
+    }
+}
+
+//// GroupElement operations.
+
+impl<C: GroupElement> Poly<C> {
+    /// Returns a polynomial with the zero element.
+    pub fn zero() -> Self {
+        Self::from(vec![C::zero()])
+    }
+
+    /// Performs polynomial addition in place.
+    pub fn add(&mut self, other: &Self) {
+        // if we have a smaller degree we should pad with zeros
+        if self.0.len() < other.0.len() {
+            self.0.resize(other.0.len(), C::zero())
+        }
+        self.0.iter_mut().zip(&other.0).for_each(|(a, b)| *a += *b)
+    }
+
+    // TODO: Some of the functions/steps below may be executed many times in practice thus cache can be
+    // used to improve efficiency (e.g., eval(i) may be called with the same index every time a partial
+    // signature from party i is verified).
+
+    /// Evaluates the polynomial at the specified value.
+    pub fn eval(&self, i: Idx) -> Eval<C> {
+        assert!(i > 0); // Never reveal the secret coefficient directly.
+        let xi = C::ScalarType::from(i.into());
+        let res = self.0.iter().rev().fold(C::zero(), |mut sum, coeff| {
+            sum = sum * xi + coeff;
+            sum
+        });
+
+        Eval {
+            index: i,
+            value: res,
+        }
+    }
+
+    /// Given at least `t` polynomial evaluations, it will recover the polynomial's
+    /// constant term
+    pub fn recover_c0(t: u32, shares: Vec<Eval<C>>) -> Result<C, FastCryptoError> {
+        let xs = Self::share_map(t, shares)?;
+
+        // Iterate over all indices and for each multiply the lagrange basis
+        // with the value of the share.
+        let mut acc = C::zero();
+        for (i, xi) in &xs {
+            let mut yi = *xi;
+            let mut num = C::ScalarType::generator();
+            let mut den = C::ScalarType::generator();
+
+            for j in xs.keys() {
+                if i == j {
+                    continue;
+                };
+                // xj - 0
+                num = num * C::ScalarType::from(*j as u64);
+                // 1 / (xj - xi)
+                let tmp = C::ScalarType::from(*j as u64) - C::ScalarType::from(*i as u64);
+                den = den * tmp;
+            }
+            // Next line is safe since i != j.
+            let inv = C::ScalarType::generator() / den;
+            num = num * inv;
```

Comment:
you can omit these operations and update `acc += yi * num * inv` directly, no?
Actually you could even avoid setting the temp `yi` and `inv` values and do `acc += *xi * num * C::ScalarType::generator() / den`